XA 模式一个分支超时，会把IDLE 连接放回连接池，导致其他业务执行异常

<!-- Please do not use this issue template to report security vulnerabilities but refer to our [security policy](https://github.com/seata/seata/security/policy). -->

- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
开启 XA 全局事务后，如果其中一条 SQL 执行超时，会导致该分支回滚失败，并且以  `IDLE ` 的状态放回连接池，后续再执行 SQL 会报错。


### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. 开启全局事务
2. 执行分支事务 A，对记录 a 修改
3. 执行分支事务 B，同样对记录 a 进行修改，但是由于 A 事务未完成，死锁导致阻塞。
4. TC 检测到全局事务超时，让各分支事务回滚：事务 A 回滚成功，事务 B 由于还在执行中，回滚失败（不断刷回滚失败日志: `NO_TA` 错误，因为 B 事务还没 `XA prepare`）
5. 由于事务 A 已回滚，事务 B 继续执行，然后尝试 `commit`，先 `XA end` 再向 TC 查询状态，由于分布式事务已经没了，报错
6. 报错之后尝试 `rollback`，rollback 时再次执行 `XA end`，mysql 返回错误 `XAER_RMFAIL: The command cannot be executed when global transaction is in the IDLE state`，回滚失败
7. 至此，当前连接处于 `IDLE` 状态，没有 `XA rollback` 成功

根本问题时因为 commit 的时候已经 `XA end` 过一次了，rollback 的时候又想 `XA end` 一次，中断了 rollback 过程。

### Ⅴ. Anything else we need to know?
建议方案:
1. 执行完 sql 后，commit 之前先向 TC 确认一次事务状态，如果事务已经过期了，直接执行 rollback
2. rollback 时执行 `XA end` 之后如果 sql 报错是 `IDLE` 错误，忽略该错误继续 `XA rollback`


如果认定这是一个 bug，我可以尝试提交修改的 PR 😀

### Ⅵ. Environment:

- JDK version(e.g. `java -version`): 11
- Seata client/server version:  1.8.0
- Database version: 8.0.29

> 所以在commit的时候为什么会走到rollback？

commit() -> end() -> termination() -> SQLException
``` java
private void termination(String xaBranchXid) throws SQLException {
        // if it is not empty, the resource will hang and need to be terminated early
        BranchStatus branchStatus = BaseDataSourceResource.getBranchStatus(xaBranchXid);
        if (branchStatus != null) {
            releaseIfNecessary();
            throw new SQLException("failed xa branch " + xid
                    + " the global transaction has finish, branch status: " + branchStatus.getCode());
        }
    }
```


我创建了一个复现的 [项目](https://github.com/tanyaofei/seata-xa-exception-sample)
1. 数据库地址 127.0.0.1:3306, 账号 root, 密码 123456，表初始化 sql 位于 `resources/database.sql`。
2. 程序启动后访问 http://127.0.0.1:8080/user。

**访问后控制台会不断地刷回滚报错日志，请耐心等待直至抛出异常**


以下是控制台报错日志
```
org.springframework.transaction.TransactionSystemException: JDBC commit failed; nested exception is java.sql.SQLException: failed xa branch 192.168.6.132:8091:5161638839849824278 the global transaction has finish, branch status: 8
	at org.springframework.jdbc.datasource.DataSourceTransactionManager.translateException(DataSourceTransactionManager.java:435)
	at org.springframework.jdbc.support.JdbcTransactionManager.translateException(JdbcTransactionManager.java:188)
	at org.springframework.jdbc.datasource.DataSourceTransactionManager.doCommit(DataSourceTransactionManager.java:336)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:743)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:711)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:152)
	at com.example.UserServiceImpl.updateUser(UserServiceImpl.java:40)
	at com.example.UserServiceImpl$$FastClassBySpringCGLIB$$4998c799.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at io.seata.spring.annotation.GlobalTransactionalInterceptor$2.execute(GlobalTransactionalInterceptor.java:208)
	at io.seata.tm.api.TransactionalTemplate.execute(TransactionalTemplate.java:128)
	at io.seata.spring.annotation.GlobalTransactionalInterceptor.handleGlobalTransaction(GlobalTransactionalInterceptor.java:205)
	at io.seata.spring.annotation.GlobalTransactionalInterceptor.invoke(GlobalTransactionalInterceptor.java:175)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
	at com.example.UserServiceImpl$$EnhancerBySpringCGLIB$$cae137cc.updateUser(<generated>)
	at com.example.UserController.getUsers(UserController.java:24)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1070)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:655)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:890)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1789)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.sql.SQLException: failed xa branch 192.168.6.132:8091:5161638839849824278 the global transaction has finish, branch status: 8
	at io.seata.rm.datasource.xa.ConnectionProxyXA.termination(ConnectionProxyXA.java:349)
	at io.seata.rm.datasource.xa.ConnectionProxyXA.termination(ConnectionProxyXA.java:340)
	at io.seata.rm.datasource.xa.ConnectionProxyXA.end(ConnectionProxyXA.java:273)
	at io.seata.rm.datasource.xa.ConnectionProxyXA.commit(ConnectionProxyXA.java:209)
	at org.springframework.jdbc.datasource.DataSourceTransactionManager.doCommit(DataSourceTransactionManager.java:333)
	... 68 more
PreparedStatementCallback; uncategorized SQLException; SQL state [XAE07]; error code [1399]; XAER_RMFAIL: The command cannot be executed when global transaction is in the  IDLE state; nested exception is java.sql.SQLException: XAER_RMFAIL: The command cannot be executed when global transaction is in the  IDLE state
PreparedStatementCallback; uncategorized SQLException; SQL state [XAE07]; error code [1399]; XAER_RMFAIL: The command cannot be executed when global transaction is in the  IDLE state; nested exception is java.sql.SQLException: XAER_RMFAIL: The command cannot be executed when global transaction is in the  IDLE state
PreparedStatementCallback; uncategorized SQLException; SQL state [XAE07]; error code [1399]; XAER_RMFAIL: The command cannot be executed when global transaction is in the  IDLE state; nested exception is java.sql.SQLException: XAER_RMFAIL: The command cannot be executed when global transaction is in the  IDLE state
```

--- a/changes/en-us/2.x.md
+++ b/changes/en-us/2.x.md
@@ -36,6 +36,7 @@ Add changes here for all PR submitted to the 2.x branch.
 - [[#6385](https://github.com/apache/incubator-seata/pull/6385)] fix the bug where Role.participant does not execute hooks but clears them.
 - [[#6465](https://github.com/apache/incubator-seata/pull/6465)] fix(6257): fix saga mode replay context lost start in 2.x
 - [[#6469](https://github.com/apache/incubator-seata/pull/6469)] fix Error in insert sql of [lock_table] data table to sqlserver database
+- [[#6492](https://github.com/apache/incubator-seata/pull/6492)] fix XA did not rollback but close when executing a long-running SQL(or deadlock SQL)
 
 ### optimize:
 - [[#6031](https://github.com/apache/incubator-seata/pull/6031)] add a check for the existence of the undolog table
@@ -186,5 +187,6 @@ Thanks to these contributors for their code commits. Please report an unintended
 - [yixia](https://github.com/wt-better)
 - [MikhailNavitski](https://github.com/MikhailNavitski)
 - [deung](https://github.com/deung)
+- [tanyaofei](https://github.com/tanyaofei)
 
 Also, we receive many valuable issues, questions and advices from our community. Thanks for you all.

--- a/changes/zh-cn/2.x.md
+++ b/changes/zh-cn/2.x.md
@@ -36,6 +36,7 @@
 - [[#6385](https://github.com/apache/incubator-seata/pull/6385)] 修复Role.Participant不执行hook但会清理的问题
 - [[#6465](https://github.com/apache/incubator-seata/pull/6465)] 修复2.0下saga模式的context replay丢失start问题
 - [[#6469](https://github.com/apache/incubator-seata/pull/6469)] 修复在sqlserver数据库下[lock_table]数据表的插入操作sql中存在的错误
+- [[#6492](https://github.com/apache/incubator-seata/pull/6492)] 修复XA执行长时间SQL(或死锁SQL)没有完成回滚就释放连接
 
 
 ### optimize:
@@ -185,5 +186,6 @@
 - [yixia](https://github.com/wt-better)
 - [MikhailNavitski](https://github.com/MikhailNavitski)
 - [deung](https://github.com/deung)
+- [tanyaofei](https://github.com/tanyaofei)
 
 同时，我们收到了社区反馈的很多有价值的issue和建议，非常感谢大家。

--- a/rm-datasource/src/main/java/org/apache/seata/rm/datasource/xa/ConnectionProxyXA.java
+++ b/rm-datasource/src/main/java/org/apache/seata/rm/datasource/xa/ConnectionProxyXA.java
@@ -55,6 +55,8 @@ public class ConnectionProxyXA extends AbstractConnectionProxyXA implements Hold
 
     private volatile boolean xaActive = false;
 
+    private volatile boolean xaEnded = false;
+
     private volatile boolean kept = false;
 
     private volatile boolean rollBacked = false;
@@ -108,6 +110,13 @@ public class ConnectionProxyXA extends AbstractConnectionProxyXA implements Hold
                     resource.release(xaBranchXid, this);
                 }
             }
+        }
+    }
+
+    private void xaEnd(XAXid xaXid, int flags) throws XAException {
+        if (!xaEnded) {
+            xaResource.end(xaXid, flags);
+            xaEnded = true;
         }
     }
 
@@ -206,7 +215,14 @@ public class ConnectionProxyXA extends AbstractConnectionProxyXA implements Hold
         }
         try {
             // XA End: Success
-            end(XAResource.TMSUCCESS);
+            try {
+                end(XAResource.TMSUCCESS);
+            } catch (SQLException sqle) {
+                // Rollback immediately before the XA Branch Context is deleted.
+                String xaBranchXid = this.xaBranchXid.toString();
+                rollback();
+                throw new SQLException("Branch " + xaBranchXid + " was rollbacked on commiting since " + sqle.getMessage(), SQLSTATE_XA_NOT_END, sqle);
+            }
             long now = System.currentTimeMillis();
             checkTimeout(now);
             setPrepareTime(now);
@@ -234,7 +250,7 @@ public class ConnectionProxyXA extends AbstractConnectionProxyXA implements Hold
         try {
             if (!rollBacked) {
                 // XA End: Fail
-                xaResource.end(this.xaBranchXid, XAResource.TMFAIL);
+                xaEnd(xaBranchXid, XAResource.TMFAIL);
                 xaRollback(xaBranchXid);
             }
             // Branch Report to TC
@@ -269,7 +285,7 @@ public class ConnectionProxyXA extends AbstractConnectionProxyXA implements Hold
     }
 
     private synchronized void end(int flags) throws XAException, SQLException {
-        xaResource.end(xaBranchXid, flags);
+        xaEnd(xaBranchXid, flags);
         termination();
     }
 
@@ -307,6 +323,7 @@ public class ConnectionProxyXA extends AbstractConnectionProxyXA implements Hold
         }
         // Force close the physical connection
         physicalConn.close();
+        xaEnded = false;
         rollBacked = false;
         cleanXABranchContext();
         originalConnection.close();

--- a/rm-datasource/src/main/java/org/apache/seata/rm/datasource/xa/ConnectionProxyXA.java
+++ b/rm-datasource/src/main/java/org/apache/seata/rm/datasource/xa/ConnectionProxyXA.java
@@ -221,7 +221,7 @@ public class ConnectionProxyXA extends AbstractConnectionProxyXA implements Hold
                 // Rollback immediately before the XA Branch Context is deleted.
                 String xaBranchXid = this.xaBranchXid.toString();
                 rollback();
-                throw new SQLException("Branch " + xaBranchXid + " was rollbacked on commiting since " + sqle.getMessage(), SQLSTATE_XA_NOT_END, sqle);
+                throw new SQLException("Branch " + xaBranchXid + " was rollbacked on committing since " + sqle.getMessage(), SQLSTATE_XA_NOT_END, sqle);
             }
             long now = System.currentTimeMillis();
             checkTimeout(now);

